{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Creating-the-Gene-Table\" data-toc-modified-id=\"Creating-the-Gene-Table-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Creating the Gene Table</a></span><ul class=\"toc-item\"><li><span><a href=\"#Get-information-from-GFF-files\" data-toc-modified-id=\"Get-information-from-GFF-files-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Get information from GFF files</a></span></li><li><span><a href=\"#(Optional)-KEGG-and-COGs\" data-toc-modified-id=\"(Optional)-KEGG-and-COGs-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>(Optional) KEGG and COGs</a></span><ul class=\"toc-item\"><li><span><a href=\"#Generate-nucleotide-fasta-files-for-CDS\" data-toc-modified-id=\"Generate-nucleotide-fasta-files-for-CDS-1.2.1\"><span class=\"toc-item-num\">1.2.1&nbsp;&nbsp;</span>Generate nucleotide fasta files for CDS</a></span></li><li><span><a href=\"#Run-EggNOG-Mapper\" data-toc-modified-id=\"Run-EggNOG-Mapper-1.2.2\"><span class=\"toc-item-num\">1.2.2&nbsp;&nbsp;</span>Run EggNOG Mapper</a></span></li><li><span><a href=\"#Get-KEGG-IDs\" data-toc-modified-id=\"Get-KEGG-IDs-1.2.3\"><span class=\"toc-item-num\">1.2.3&nbsp;&nbsp;</span>Get KEGG IDs</a></span></li><li><span><a href=\"#Save-KEGG-information\" data-toc-modified-id=\"Save-KEGG-information-1.2.4\"><span class=\"toc-item-num\">1.2.4&nbsp;&nbsp;</span>Save KEGG information</a></span></li><li><span><a href=\"#Save-COGs-to-annotation-dataframe\" data-toc-modified-id=\"Save-COGs-to-annotation-dataframe-1.2.5\"><span class=\"toc-item-num\">1.2.5&nbsp;&nbsp;</span>Save COGs to annotation dataframe</a></span></li></ul></li><li><span><a href=\"#Uniprot-ID-mapping\" data-toc-modified-id=\"Uniprot-ID-mapping-1.3\"><span class=\"toc-item-num\">1.3&nbsp;&nbsp;</span>Uniprot ID mapping</a></span></li><li><span><a href=\"#Add-Biocyc-Operon-information\" data-toc-modified-id=\"Add-Biocyc-Operon-information-1.4\"><span class=\"toc-item-num\">1.4&nbsp;&nbsp;</span>Add Biocyc Operon information</a></span><ul class=\"toc-item\"><li><span><a href=\"#Assign-unique-IDs-to-operons\" data-toc-modified-id=\"Assign-unique-IDs-to-operons-1.4.1\"><span class=\"toc-item-num\">1.4.1&nbsp;&nbsp;</span>Assign unique IDs to operons</a></span></li></ul></li><li><span><a href=\"#Clean-up-and-save-annotation\" data-toc-modified-id=\"Clean-up-and-save-annotation-1.5\"><span class=\"toc-item-num\">1.5&nbsp;&nbsp;</span>Clean up and save annotation</a></span></li><li><span><a href=\"#Final-statistics\" data-toc-modified-id=\"Final-statistics-1.6\"><span class=\"toc-item-num\">1.6&nbsp;&nbsp;</span>Final statistics</a></span></li><li><span><a href=\"#Fill-missing-values\" data-toc-modified-id=\"Fill-missing-values-1.7\"><span class=\"toc-item-num\">1.7&nbsp;&nbsp;</span>Fill missing values</a></span></li><li><span><a href=\"#GO-Annotations\" data-toc-modified-id=\"GO-Annotations-1.8\"><span class=\"toc-item-num\">1.8&nbsp;&nbsp;</span>GO Annotations</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating the Gene Table\n",
    "This notebook is copied from the [Pymodulon GitHub repository](https://github.com/SBRG/pymodulon/blob/master/docs/tutorials/creating_the_gene_table.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymodulon.gene_util import *\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get information from GFF files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, download the FASTA and GFF files for your organism and its plasmids from NCBI."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enter the location of all your GFF files here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_files = [os.path.join('..','data','external','genome.gff3')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following cell will convert all the GFF files into a single Pandas DataFrame for easy manipulation. Pseudogenes have multiple rows in a GFF file (one for each fragment), but only the first fragment will be kept."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>ncbi_protein</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_1768</th>\n",
       "      <td>CP006891.1</td>\n",
       "      <td>32.0</td>\n",
       "      <td>841.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Phage capsid scaffolding protein (GPO) serine ...</td>\n",
       "      <td>AHB11032.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1853</th>\n",
       "      <td>CP006893.1</td>\n",
       "      <td>80.0</td>\n",
       "      <td>2122.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>outer membrane receptor protein</td>\n",
       "      <td>AHB11110.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1912</th>\n",
       "      <td>CP006895.1</td>\n",
       "      <td>450.0</td>\n",
       "      <td>680.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11159.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1913</th>\n",
       "      <td>CP006895.1</td>\n",
       "      <td>677.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prophage maintenance system killer protein</td>\n",
       "      <td>AHB11160.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1769</th>\n",
       "      <td>CP006891.1</td>\n",
       "      <td>974.0</td>\n",
       "      <td>2749.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11033.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession  start     end strand gene_name old_locus_tag  \\\n",
       "locus_tag                                                             \n",
       "ZCP4_1768  CP006891.1   32.0   841.0      -      None          None   \n",
       "ZCP4_1853  CP006893.1   80.0  2122.0      -      None          None   \n",
       "ZCP4_1912  CP006895.1  450.0   680.0      +      None          None   \n",
       "ZCP4_1913  CP006895.1  677.0  1057.0      +      None          None   \n",
       "ZCP4_1769  CP006891.1  974.0  2749.0      +      None          None   \n",
       "\n",
       "                                                gene_product ncbi_protein  \n",
       "locus_tag                                                                  \n",
       "ZCP4_1768  Phage capsid scaffolding protein (GPO) serine ...   AHB11032.1  \n",
       "ZCP4_1853                    outer membrane receptor protein   AHB11110.1  \n",
       "ZCP4_1912                               hypothetical protein   AHB11159.1  \n",
       "ZCP4_1913         prophage maintenance system killer protein   AHB11160.1  \n",
       "ZCP4_1769                               hypothetical protein   AHB11033.1  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "keep_cols = ['accession','start','end','strand','gene_name','old_locus_tag','gene_product','ncbi_protein']\n",
    "\n",
    "DF_annot = gff2pandas(gff_files,index='locus_tag')\n",
    "DF_annot = DF_annot[keep_cols]\n",
    "\n",
    "DF_annot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To ensure that the gene index used is identical to the expression matrix, load in your data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SRX14953908</th>\n",
       "      <th>SRX14953909</th>\n",
       "      <th>SRX14953910</th>\n",
       "      <th>SRX14953911</th>\n",
       "      <th>SRX14953912</th>\n",
       "      <th>SRX14953913</th>\n",
       "      <th>SRX14953914</th>\n",
       "      <th>SRX14953915</th>\n",
       "      <th>SRX14953916</th>\n",
       "      <th>SRX14953917</th>\n",
       "      <th>...</th>\n",
       "      <th>Kiley_B_WT_-O2_Ctrl_B</th>\n",
       "      <th>Kiley_B_WT_-O2_Ctrl_C</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_A</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_B</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_C</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_DIP_A</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_DIP_B</th>\n",
       "      <th>Kiley_B_WT_-O2_NoFe_DIP_C</th>\n",
       "      <th>Kiley_B_WT_-O2_Peroxide_A</th>\n",
       "      <th>Kiley_B_WT_-O2_Peroxide_C</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_0001</th>\n",
       "      <td>5.776818</td>\n",
       "      <td>5.914021</td>\n",
       "      <td>5.649816</td>\n",
       "      <td>5.770222</td>\n",
       "      <td>5.634935</td>\n",
       "      <td>7.319695</td>\n",
       "      <td>7.463173</td>\n",
       "      <td>7.188563</td>\n",
       "      <td>7.053904</td>\n",
       "      <td>5.838514</td>\n",
       "      <td>...</td>\n",
       "      <td>6.753442</td>\n",
       "      <td>6.817656</td>\n",
       "      <td>6.721457</td>\n",
       "      <td>6.920074</td>\n",
       "      <td>7.001997</td>\n",
       "      <td>6.937601</td>\n",
       "      <td>6.924576</td>\n",
       "      <td>6.861227</td>\n",
       "      <td>3.711552</td>\n",
       "      <td>5.354796</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0002</th>\n",
       "      <td>6.729829</td>\n",
       "      <td>7.189463</td>\n",
       "      <td>6.933535</td>\n",
       "      <td>6.985748</td>\n",
       "      <td>6.964886</td>\n",
       "      <td>6.967124</td>\n",
       "      <td>7.090734</td>\n",
       "      <td>6.816982</td>\n",
       "      <td>6.805465</td>\n",
       "      <td>7.045432</td>\n",
       "      <td>...</td>\n",
       "      <td>6.776425</td>\n",
       "      <td>7.010168</td>\n",
       "      <td>6.752740</td>\n",
       "      <td>6.829927</td>\n",
       "      <td>6.808803</td>\n",
       "      <td>6.901704</td>\n",
       "      <td>6.883673</td>\n",
       "      <td>6.843651</td>\n",
       "      <td>3.466233</td>\n",
       "      <td>5.262928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0003</th>\n",
       "      <td>12.218703</td>\n",
       "      <td>12.306204</td>\n",
       "      <td>12.578060</td>\n",
       "      <td>12.411400</td>\n",
       "      <td>12.231102</td>\n",
       "      <td>8.677390</td>\n",
       "      <td>8.529919</td>\n",
       "      <td>8.539999</td>\n",
       "      <td>8.463958</td>\n",
       "      <td>12.439551</td>\n",
       "      <td>...</td>\n",
       "      <td>12.279739</td>\n",
       "      <td>11.998652</td>\n",
       "      <td>12.358939</td>\n",
       "      <td>12.264778</td>\n",
       "      <td>12.285701</td>\n",
       "      <td>12.185583</td>\n",
       "      <td>12.215878</td>\n",
       "      <td>12.003532</td>\n",
       "      <td>12.259624</td>\n",
       "      <td>12.380901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0004</th>\n",
       "      <td>9.114917</td>\n",
       "      <td>8.863234</td>\n",
       "      <td>8.965472</td>\n",
       "      <td>8.971342</td>\n",
       "      <td>9.270446</td>\n",
       "      <td>8.902531</td>\n",
       "      <td>8.916778</td>\n",
       "      <td>9.007271</td>\n",
       "      <td>8.876492</td>\n",
       "      <td>9.179108</td>\n",
       "      <td>...</td>\n",
       "      <td>8.102858</td>\n",
       "      <td>8.336005</td>\n",
       "      <td>8.253104</td>\n",
       "      <td>8.416695</td>\n",
       "      <td>8.319913</td>\n",
       "      <td>8.498210</td>\n",
       "      <td>8.605744</td>\n",
       "      <td>8.557059</td>\n",
       "      <td>9.386926</td>\n",
       "      <td>9.178899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0005</th>\n",
       "      <td>10.756126</td>\n",
       "      <td>10.485917</td>\n",
       "      <td>10.354395</td>\n",
       "      <td>10.437860</td>\n",
       "      <td>10.640383</td>\n",
       "      <td>14.226580</td>\n",
       "      <td>14.203014</td>\n",
       "      <td>13.921314</td>\n",
       "      <td>14.050698</td>\n",
       "      <td>10.794631</td>\n",
       "      <td>...</td>\n",
       "      <td>9.568082</td>\n",
       "      <td>9.261161</td>\n",
       "      <td>9.387126</td>\n",
       "      <td>9.473256</td>\n",
       "      <td>9.546165</td>\n",
       "      <td>9.413043</td>\n",
       "      <td>9.381610</td>\n",
       "      <td>9.326745</td>\n",
       "      <td>10.637753</td>\n",
       "      <td>11.172003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 312 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           SRX14953908  SRX14953909  SRX14953910  SRX14953911  SRX14953912  \\\n",
       "Geneid                                                                       \n",
       "ZCP4_0001     5.776818     5.914021     5.649816     5.770222     5.634935   \n",
       "ZCP4_0002     6.729829     7.189463     6.933535     6.985748     6.964886   \n",
       "ZCP4_0003    12.218703    12.306204    12.578060    12.411400    12.231102   \n",
       "ZCP4_0004     9.114917     8.863234     8.965472     8.971342     9.270446   \n",
       "ZCP4_0005    10.756126    10.485917    10.354395    10.437860    10.640383   \n",
       "\n",
       "           SRX14953913  SRX14953914  SRX14953915  SRX14953916  SRX14953917  \\\n",
       "Geneid                                                                       \n",
       "ZCP4_0001     7.319695     7.463173     7.188563     7.053904     5.838514   \n",
       "ZCP4_0002     6.967124     7.090734     6.816982     6.805465     7.045432   \n",
       "ZCP4_0003     8.677390     8.529919     8.539999     8.463958    12.439551   \n",
       "ZCP4_0004     8.902531     8.916778     9.007271     8.876492     9.179108   \n",
       "ZCP4_0005    14.226580    14.203014    13.921314    14.050698    10.794631   \n",
       "\n",
       "           ...  Kiley_B_WT_-O2_Ctrl_B  Kiley_B_WT_-O2_Ctrl_C  \\\n",
       "Geneid     ...                                                 \n",
       "ZCP4_0001  ...               6.753442               6.817656   \n",
       "ZCP4_0002  ...               6.776425               7.010168   \n",
       "ZCP4_0003  ...              12.279739              11.998652   \n",
       "ZCP4_0004  ...               8.102858               8.336005   \n",
       "ZCP4_0005  ...               9.568082               9.261161   \n",
       "\n",
       "           Kiley_B_WT_-O2_NoFe_A  Kiley_B_WT_-O2_NoFe_B  \\\n",
       "Geneid                                                    \n",
       "ZCP4_0001               6.721457               6.920074   \n",
       "ZCP4_0002               6.752740               6.829927   \n",
       "ZCP4_0003              12.358939              12.264778   \n",
       "ZCP4_0004               8.253104               8.416695   \n",
       "ZCP4_0005               9.387126               9.473256   \n",
       "\n",
       "           Kiley_B_WT_-O2_NoFe_C  Kiley_B_WT_-O2_NoFe_DIP_A  \\\n",
       "Geneid                                                        \n",
       "ZCP4_0001               7.001997                   6.937601   \n",
       "ZCP4_0002               6.808803                   6.901704   \n",
       "ZCP4_0003              12.285701                  12.185583   \n",
       "ZCP4_0004               8.319913                   8.498210   \n",
       "ZCP4_0005               9.546165                   9.413043   \n",
       "\n",
       "           Kiley_B_WT_-O2_NoFe_DIP_B  Kiley_B_WT_-O2_NoFe_DIP_C  \\\n",
       "Geneid                                                            \n",
       "ZCP4_0001                   6.924576                   6.861227   \n",
       "ZCP4_0002                   6.883673                   6.843651   \n",
       "ZCP4_0003                  12.215878                  12.003532   \n",
       "ZCP4_0004                   8.605744                   8.557059   \n",
       "ZCP4_0005                   9.381610                   9.326745   \n",
       "\n",
       "           Kiley_B_WT_-O2_Peroxide_A  Kiley_B_WT_-O2_Peroxide_C  \n",
       "Geneid                                                           \n",
       "ZCP4_0001                   3.711552                   5.354796  \n",
       "ZCP4_0002                   3.466233                   5.262928  \n",
       "ZCP4_0003                  12.259624                  12.380901  \n",
       "ZCP4_0004                   9.386926                   9.178899  \n",
       "ZCP4_0005                  10.637753                  11.172003  \n",
       "\n",
       "[5 rows x 312 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log_tpm_file = os.path.join('..','data','processed_data','log_tpm.csv')\n",
    "DF_log_tpm = pd.read_csv(log_tpm_file,index_col=0)\n",
    "DF_log_tpm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check that the genes are the same in the expression dataset as in the annotation dataframe. Mismatched genes are listed below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ZCP4_0001', 'ZCP4_0002', 'ZCP4_0003', 'ZCP4_0004'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(DF_log_tpm.index) - set(DF_annot.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (Optional) KEGG and COGs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate nucleotide fasta files for CDS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Enter the location of all your fasta files here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_files = [os.path.join('..','data','external','genome.fasta')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code generates CDS files using your FASTA and GFF3 files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "cds_list = []\n",
    "for fasta in fasta_files:\n",
    "    for seq in SeqIO.parse(fasta, 'fasta'):  # Use SeqIO.parse to handle multiple sequences\n",
    "\n",
    "        # Get gene information for genes in this fasta file\n",
    "        df_genes = DF_annot[DF_annot.accession == seq.id]\n",
    "\n",
    "        for i, row in df_genes.iterrows():\n",
    "            cds = seq[int(row.start-1):int(row.end)]\n",
    "            if row.strand == '-':\n",
    "                cds = seq[int(row.start-1):int(row.end)].reverse_complement()\n",
    "            cds.id = row.name\n",
    "            cds.description = row.gene_name if pd.notnull(row.gene_name) else row.name\n",
    "            cds_list.append(cds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save the CDS file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1836"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_file = os.path.join('..','data','external','CDS.fna')\n",
    "SeqIO.write(cds_list, cds_file, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Run EggNOG Mapper\n",
    "1. Go to http://eggnog-mapper.embl.de/.\n",
    "1. Upload the CDS.fna file from your organism directory (within the sequence_files folder)\n",
    "1. Make sure to limit the taxonomy to the correct level\n",
    "1. After the job is submitted, you must follow the link in your email to run the job.\n",
    "1. Once the job completes (after ~4 hrs), download the annotations file.\n",
    "1. Save the annotation file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get KEGG IDs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once you have the EggNOG annotations, load the annotation file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnog_file = os.path.join('..','data','external','eggNOG_annotations.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seed eggNOG ortholog</th>\n",
       "      <th>seed ortholog evalue</th>\n",
       "      <th>seed ortholog score</th>\n",
       "      <th>eggNOG OGs</th>\n",
       "      <th>max_annot_lvl</th>\n",
       "      <th>COG</th>\n",
       "      <th>Description</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC number</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_pathway</th>\n",
       "      <th>KEGG_module</th>\n",
       "      <th>KEGG_reaction</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG Reaction</th>\n",
       "      <th>PFAMs</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_0005</th>\n",
       "      <td>264203.ZMO1352</td>\n",
       "      <td>2.180000e-159</td>\n",
       "      <td>448.0</td>\n",
       "      <td>COG0670@1|root,COG0670@2|Bacteria,1MU69@1224|P...</td>\n",
       "      <td>204457|Sphingomonadales</td>\n",
       "      <td>S</td>\n",
       "      <td>Belongs to the BI1 family</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko:K06890</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Bax1-I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0006</th>\n",
       "      <td>264203.ZMO1351</td>\n",
       "      <td>3.880000e-240</td>\n",
       "      <td>660.0</td>\n",
       "      <td>COG0412@1|root,COG0412@2|Bacteria,1MW7S@1224|P...</td>\n",
       "      <td>204457|Sphingomonadales</td>\n",
       "      <td>Q</td>\n",
       "      <td>dienelactone hydrolase</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>3.1.1.45</td>\n",
       "      <td>ko:K01061</td>\n",
       "      <td>ko00361,ko00364,ko00623,ko01100,ko01110,ko0112...</td>\n",
       "      <td>-</td>\n",
       "      <td>R03893,R05510,R05511,R06835,R06838,R08120,R081...</td>\n",
       "      <td>RC01018,RC01906,RC01907,RC02441,RC02467,RC0246...</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>DLH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0007</th>\n",
       "      <td>264203.ZMO1350</td>\n",
       "      <td>3.280000e-180</td>\n",
       "      <td>501.0</td>\n",
       "      <td>COG1058@1|root,COG1058@2|Bacteria,1MVG6@1224|P...</td>\n",
       "      <td>204457|Sphingomonadales</td>\n",
       "      <td>S</td>\n",
       "      <td>nucleotide-utilizing enzyme related to molybdo...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MoCF_biosynth</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0008</th>\n",
       "      <td>264203.ZMO1348</td>\n",
       "      <td>4.050000e-206</td>\n",
       "      <td>569.0</td>\n",
       "      <td>COG0024@1|root,COG0024@2|Bacteria,1MU99@1224|P...</td>\n",
       "      <td>204457|Sphingomonadales</td>\n",
       "      <td>E</td>\n",
       "      <td>Methionine aminopeptidase</td>\n",
       "      <td>map</td>\n",
       "      <td>-</td>\n",
       "      <td>3.4.11.18</td>\n",
       "      <td>ko:K01265</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ko00000,ko01000,ko01002</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Peptidase_M24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0009</th>\n",
       "      <td>264203.ZMO1347</td>\n",
       "      <td>1.080000e-243</td>\n",
       "      <td>669.0</td>\n",
       "      <td>COG2008@1|root,COG2008@2|Bacteria,1MWCR@1224|P...</td>\n",
       "      <td>204457|Sphingomonadales</td>\n",
       "      <td>E</td>\n",
       "      <td>Threonine aldolase</td>\n",
       "      <td>ltaA</td>\n",
       "      <td>-</td>\n",
       "      <td>4.1.2.48</td>\n",
       "      <td>ko:K01620</td>\n",
       "      <td>ko00260,ko01100,ko01110,ko01120,ko01130,ko0123...</td>\n",
       "      <td>-</td>\n",
       "      <td>R00751,R06171</td>\n",
       "      <td>RC00312,RC00372</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>Beta_elim_lyase</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seed eggNOG ortholog  seed ortholog evalue  seed ortholog score  \\\n",
       "locus_tag                                                                   \n",
       "ZCP4_0005       264203.ZMO1352         2.180000e-159                448.0   \n",
       "ZCP4_0006       264203.ZMO1351         3.880000e-240                660.0   \n",
       "ZCP4_0007       264203.ZMO1350         3.280000e-180                501.0   \n",
       "ZCP4_0008       264203.ZMO1348         4.050000e-206                569.0   \n",
       "ZCP4_0009       264203.ZMO1347         1.080000e-243                669.0   \n",
       "\n",
       "                                                  eggNOG OGs  \\\n",
       "locus_tag                                                      \n",
       "ZCP4_0005  COG0670@1|root,COG0670@2|Bacteria,1MU69@1224|P...   \n",
       "ZCP4_0006  COG0412@1|root,COG0412@2|Bacteria,1MW7S@1224|P...   \n",
       "ZCP4_0007  COG1058@1|root,COG1058@2|Bacteria,1MVG6@1224|P...   \n",
       "ZCP4_0008  COG0024@1|root,COG0024@2|Bacteria,1MU99@1224|P...   \n",
       "ZCP4_0009  COG2008@1|root,COG2008@2|Bacteria,1MWCR@1224|P...   \n",
       "\n",
       "                     max_annot_lvl COG  \\\n",
       "locus_tag                                \n",
       "ZCP4_0005  204457|Sphingomonadales   S   \n",
       "ZCP4_0006  204457|Sphingomonadales   Q   \n",
       "ZCP4_0007  204457|Sphingomonadales   S   \n",
       "ZCP4_0008  204457|Sphingomonadales   E   \n",
       "ZCP4_0009  204457|Sphingomonadales   E   \n",
       "\n",
       "                                                 Description Preferred_name  \\\n",
       "locus_tag                                                                     \n",
       "ZCP4_0005                          Belongs to the BI1 family              -   \n",
       "ZCP4_0006                             dienelactone hydrolase              -   \n",
       "ZCP4_0007  nucleotide-utilizing enzyme related to molybdo...              -   \n",
       "ZCP4_0008                          Methionine aminopeptidase            map   \n",
       "ZCP4_0009                                 Threonine aldolase           ltaA   \n",
       "\n",
       "          GOs  EC number    KEGG_ko  \\\n",
       "locus_tag                             \n",
       "ZCP4_0005   -          -  ko:K06890   \n",
       "ZCP4_0006   -   3.1.1.45  ko:K01061   \n",
       "ZCP4_0007   -          -          -   \n",
       "ZCP4_0008   -  3.4.11.18  ko:K01265   \n",
       "ZCP4_0009   -   4.1.2.48  ko:K01620   \n",
       "\n",
       "                                                KEGG_pathway KEGG_module  \\\n",
       "locus_tag                                                                  \n",
       "ZCP4_0005                                                  -           -   \n",
       "ZCP4_0006  ko00361,ko00364,ko00623,ko01100,ko01110,ko0112...           -   \n",
       "ZCP4_0007                                                  -           -   \n",
       "ZCP4_0008                                                  -           -   \n",
       "ZCP4_0009  ko00260,ko01100,ko01110,ko01120,ko01130,ko0123...           -   \n",
       "\n",
       "                                               KEGG_reaction  \\\n",
       "locus_tag                                                      \n",
       "ZCP4_0005                                                  -   \n",
       "ZCP4_0006  R03893,R05510,R05511,R06835,R06838,R08120,R081...   \n",
       "ZCP4_0007                                                  -   \n",
       "ZCP4_0008                                                  -   \n",
       "ZCP4_0009                                      R00751,R06171   \n",
       "\n",
       "                                                 KEGG_rclass  \\\n",
       "locus_tag                                                      \n",
       "ZCP4_0005                                                  -   \n",
       "ZCP4_0006  RC01018,RC01906,RC01907,RC02441,RC02467,RC0246...   \n",
       "ZCP4_0007                                                  -   \n",
       "ZCP4_0008                                                  -   \n",
       "ZCP4_0009                                    RC00312,RC00372   \n",
       "\n",
       "                             BRITE KEGG_TC CAZy BiGG Reaction            PFAMs  \n",
       "locus_tag                                                                       \n",
       "ZCP4_0005                  ko00000       -    -             -           Bax1-I  \n",
       "ZCP4_0006  ko00000,ko00001,ko01000       -    -             -              DLH  \n",
       "ZCP4_0007                        -       -    -             -    MoCF_biosynth  \n",
       "ZCP4_0008  ko00000,ko01000,ko01002       -    -             -    Peptidase_M24  \n",
       "ZCP4_0009  ko00000,ko00001,ko01000       -    -             -  Beta_elim_lyase  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_eggnog = pd.read_csv(eggnog_file,sep='\\t',skiprows=5,header=None)\n",
    "eggnog_cols = ['query_name','seed eggNOG ortholog','seed ortholog evalue','seed ortholog score',\n",
    "               'eggNOG OGs','max_annot_lvl','COG','Description','Preferred_name','GOs',\n",
    "               'EC number','KEGG_ko','KEGG_pathway','KEGG_module','KEGG_reaction','KEGG_rclass',\n",
    "               'BRITE','KEGG_TC','CAZy','BiGG Reaction','PFAMs']\n",
    "\n",
    "DF_eggnog.columns = eggnog_cols\n",
    "\n",
    "# Strip last three rows as they are comments\n",
    "DF_eggnog = DF_eggnog.iloc[:-3]\n",
    "\n",
    "# Set locus tag as index\n",
    "DF_eggnog = DF_eggnog.set_index('query_name')\n",
    "DF_eggnog.index.name = 'locus_tag'\n",
    "\n",
    "DF_eggnog.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will pull the KEGG information from the eggNOG file, including orthology, pathway, module, and reactions for each gene."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>database</th>\n",
       "      <th>kegg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZCP4_0005</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ZCP4_0006</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ZCP4_0006</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ZCP4_0006</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map00623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ZCP4_0006</td>\n",
       "      <td>KEGG_pathway</td>\n",
       "      <td>map01100</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene_id      database   kegg_id\n",
       "0   ZCP4_0005  KEGG_pathway         -\n",
       "8   ZCP4_0006  KEGG_pathway  map00361\n",
       "9   ZCP4_0006  KEGG_pathway  map00364\n",
       "10  ZCP4_0006  KEGG_pathway  map00623\n",
       "11  ZCP4_0006  KEGG_pathway  map01100"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_kegg = DF_eggnog[['KEGG_pathway','KEGG_module','KEGG_reaction']]\n",
    "\n",
    "# Melt dataframe\n",
    "DF_kegg = DF_kegg.reset_index().melt(id_vars='locus_tag') \n",
    "\n",
    "# Remove null values\n",
    "DF_kegg = DF_kegg[DF_kegg.value.notnull()]\n",
    "\n",
    "# Split comma-separated values into their own rows\n",
    "list2struct = []\n",
    "for name,row in DF_kegg.iterrows():\n",
    "    for val in row.value.split(','):\n",
    "        list2struct.append([row.locus_tag,row.variable,val])\n",
    "\n",
    "DF_kegg = pd.DataFrame(list2struct,columns=['gene_id','database','kegg_id'])\n",
    "\n",
    "# Remove ko entries, as only map entries are searchable in KEGG pathway\n",
    "DF_kegg = DF_kegg[~DF_kegg.kegg_id.str.startswith('ko')]\n",
    "\n",
    "DF_kegg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save KEGG information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_kegg_ids = {}\n",
    "for index, row in DF_kegg.iterrows():\n",
    "    if 'map' not in row['kegg_id']:\n",
    "        continue\n",
    "    if row['gene_id'] not in gene_to_kegg_ids:\n",
    "        gene_to_kegg_ids.update({row['gene_id'] : set()})\n",
    "    gene_to_kegg_ids[row['gene_id']].add(row['kegg_id'])\n",
    "kegg_maps = []\n",
    "for index in DF_annot.index:\n",
    "    if index in gene_to_kegg_ids:\n",
    "        new_vals = list(gene_to_kegg_ids[index])\n",
    "        new_vals.sort()\n",
    "        kegg_maps.append(';'.join(new_vals))\n",
    "    else:\n",
    "        kegg_maps.append('')\n",
    "DF_annot['kegg_maps'] = kegg_maps\n",
    "\n",
    "\n",
    "# also save EC information\n",
    "gene_to_EC = {}\n",
    "for index, row in DF_eggnog.iterrows():\n",
    "    if '.' in row['EC number']:\n",
    "        gene_to_EC.update({index : row['EC number']})\n",
    "EC_nums = []\n",
    "for index in DF_annot.index:\n",
    "    if index in gene_to_EC:\n",
    "        EC_nums.append(gene_to_EC[index])\n",
    "    else:\n",
    "        EC_nums.append('')\n",
    "DF_annot['EC_number'] = EC_nums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_kegg.to_csv(os.path.join('..','data','external','kegg_mapping.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save COGs to annotation dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_annot['COG'] = DF_eggnog.COG\n",
    "DF_annot['EggNOG_ID'] = DF_eggnog['seed eggNOG ortholog']\n",
    "\n",
    "# Make sure COG only has one entry per gene\n",
    "DF_annot['COG'] = [item[0] if isinstance(item,str) else item for item in DF_annot['COG']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Uniprot ID mapping\n",
    "Save the NCBI protein ID to a file for uniprot mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_id_map = []\n",
    "with open(os.path.join('..','data','external','uniprot_input.txt'), \"w\") as file:\n",
    "    for item in DF_annot.EggNOG_ID.fillna('').values:\n",
    "        if '264203' not in item:\n",
    "            new_id_map.append(item)\n",
    "            continue\n",
    "        new_id_map.append('zmo:'+str(item).split('.')[1])\n",
    "        file.write('zmo:'+str(item).split('.')[1]+\"\\n\")\n",
    "    file.close()\n",
    "DF_annot['EggNOG_ID'] = new_id_map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Map the best match EggNOG matches to Zymomonas strains to Uniprot IDs\n",
    "1. Go to https://www.uniprot.org/id-mapping/.\n",
    "1. Upload the `uniprot_input.txt` file from the `external` folder\n",
    "1. For **\"From database\"**, choose **KEGG**\n",
    "1. Once the mapping completes, download the annotations file by clicking Download on the results page. Choose **\"Download all\"**, **\"Format TSV\"** \n",
    "1. Save the annotation file to the `external` folder, rename to `uniprot_mapping.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>From</th>\n",
       "      <th>Entry</th>\n",
       "      <th>Reviewed</th>\n",
       "      <th>Entry Name</th>\n",
       "      <th>Protein names</th>\n",
       "      <th>Gene Names</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zmo:ZMO1352</td>\n",
       "      <td>D2YW23</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>D2YW23_ZYMMO</td>\n",
       "      <td>Inner membrane protein YbhL</td>\n",
       "      <td>ZMO1352</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zmo:ZMO1351</td>\n",
       "      <td>Q5NMT5</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q5NMT5_ZYMMO</td>\n",
       "      <td>Carboxymethylenebutenolidase (EC 3.1.1.45)</td>\n",
       "      <td>ZMO1351</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zmo:ZMO1350</td>\n",
       "      <td>H2VFS2</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>H2VFS2_ZYMMO</td>\n",
       "      <td>Molybdopterin binding domain protein</td>\n",
       "      <td>ZMO1350</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zmo:ZMO1348</td>\n",
       "      <td>H2VFN6</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>H2VFN6_ZYMMO</td>\n",
       "      <td>Methionine aminopeptidase (MAP) (MetAP) (EC 3....</td>\n",
       "      <td>map ZMO1348</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zmo:ZMO1347</td>\n",
       "      <td>H2VFP6</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>H2VFP6_ZYMMO</td>\n",
       "      <td>Aromatic amino acid beta-eliminating lyase/thr...</td>\n",
       "      <td>ZMO1347</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>337</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>zmo:ZMO1366</td>\n",
       "      <td>Q9Z5U7</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>RS4_ZYMMO</td>\n",
       "      <td>Small ribosomal subunit protein uS4 (30S ribos...</td>\n",
       "      <td>rpsD ZMO1366</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1152</th>\n",
       "      <td>zmo:ZMO1365</td>\n",
       "      <td>Q5NMS1</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q5NMS1_ZYMMO</td>\n",
       "      <td>diguanylate cyclase (EC 2.7.7.65)</td>\n",
       "      <td>ZMO1365</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>1045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1153</th>\n",
       "      <td>zmo:ZMO1364</td>\n",
       "      <td>Q5NMS2</td>\n",
       "      <td>unreviewed</td>\n",
       "      <td>Q5NMS2_ZYMMO</td>\n",
       "      <td>Coproporphyrinogen-III oxidase (EC 1.3.98.3)</td>\n",
       "      <td>ZMO1364</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1154</th>\n",
       "      <td>zmo:ZMO1360</td>\n",
       "      <td>P06672</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>PDC_ZYMMO</td>\n",
       "      <td>Pyruvate decarboxylase (PDC) (EC 4.1.1.1)</td>\n",
       "      <td>pdc ZMO1360</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>568</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1155</th>\n",
       "      <td>zmo:ZMO1358</td>\n",
       "      <td>Q9Z5V0</td>\n",
       "      <td>reviewed</td>\n",
       "      <td>RS20_ZYMMO</td>\n",
       "      <td>Small ribosomal subunit protein bS20 (30S ribo...</td>\n",
       "      <td>rpsT ZMO1358</td>\n",
       "      <td>Zymomonas mobilis subsp. mobilis (strain ATCC ...</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1156 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             From   Entry    Reviewed    Entry Name  \\\n",
       "0     zmo:ZMO1352  D2YW23  unreviewed  D2YW23_ZYMMO   \n",
       "1     zmo:ZMO1351  Q5NMT5  unreviewed  Q5NMT5_ZYMMO   \n",
       "2     zmo:ZMO1350  H2VFS2  unreviewed  H2VFS2_ZYMMO   \n",
       "3     zmo:ZMO1348  H2VFN6  unreviewed  H2VFN6_ZYMMO   \n",
       "4     zmo:ZMO1347  H2VFP6  unreviewed  H2VFP6_ZYMMO   \n",
       "...           ...     ...         ...           ...   \n",
       "1151  zmo:ZMO1366  Q9Z5U7    reviewed     RS4_ZYMMO   \n",
       "1152  zmo:ZMO1365  Q5NMS1  unreviewed  Q5NMS1_ZYMMO   \n",
       "1153  zmo:ZMO1364  Q5NMS2  unreviewed  Q5NMS2_ZYMMO   \n",
       "1154  zmo:ZMO1360  P06672    reviewed     PDC_ZYMMO   \n",
       "1155  zmo:ZMO1358  Q9Z5V0    reviewed    RS20_ZYMMO   \n",
       "\n",
       "                                          Protein names    Gene Names  \\\n",
       "0                           Inner membrane protein YbhL       ZMO1352   \n",
       "1            Carboxymethylenebutenolidase (EC 3.1.1.45)       ZMO1351   \n",
       "2                  Molybdopterin binding domain protein       ZMO1350   \n",
       "3     Methionine aminopeptidase (MAP) (MetAP) (EC 3....   map ZMO1348   \n",
       "4     Aromatic amino acid beta-eliminating lyase/thr...       ZMO1347   \n",
       "...                                                 ...           ...   \n",
       "1151  Small ribosomal subunit protein uS4 (30S ribos...  rpsD ZMO1366   \n",
       "1152                  diguanylate cyclase (EC 2.7.7.65)       ZMO1365   \n",
       "1153       Coproporphyrinogen-III oxidase (EC 1.3.98.3)       ZMO1364   \n",
       "1154          Pyruvate decarboxylase (PDC) (EC 4.1.1.1)   pdc ZMO1360   \n",
       "1155  Small ribosomal subunit protein bS20 (30S ribo...  rpsT ZMO1358   \n",
       "\n",
       "                                               Organism  Length  \n",
       "0     Zymomonas mobilis subsp. mobilis (strain ATCC ...     242  \n",
       "1     Zymomonas mobilis subsp. mobilis (strain ATCC ...     334  \n",
       "2     Zymomonas mobilis subsp. mobilis (strain ATCC ...     253  \n",
       "3     Zymomonas mobilis subsp. mobilis (strain ATCC ...     276  \n",
       "4     Zymomonas mobilis subsp. mobilis (strain ATCC ...     337  \n",
       "...                                                 ...     ...  \n",
       "1151  Zymomonas mobilis subsp. mobilis (strain ATCC ...     204  \n",
       "1152  Zymomonas mobilis subsp. mobilis (strain ATCC ...    1045  \n",
       "1153  Zymomonas mobilis subsp. mobilis (strain ATCC ...     453  \n",
       "1154  Zymomonas mobilis subsp. mobilis (strain ATCC ...     568  \n",
       "1155  Zymomonas mobilis subsp. mobilis (strain ATCC ...      87  \n",
       "\n",
       "[1156 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv(os.path.join('..','data','external','uniprot_mapping.tsv'),sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EggNOG_ID</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot_gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>zmo:ZMO1352</td>\n",
       "      <td>D2YW23</td>\n",
       "      <td>ZMO1352</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>zmo:ZMO1351</td>\n",
       "      <td>Q5NMT5</td>\n",
       "      <td>ZMO1351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>zmo:ZMO1350</td>\n",
       "      <td>H2VFS2</td>\n",
       "      <td>ZMO1350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>zmo:ZMO1348</td>\n",
       "      <td>H2VFN6</td>\n",
       "      <td>map ZMO1348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>zmo:ZMO1347</td>\n",
       "      <td>H2VFP6</td>\n",
       "      <td>ZMO1347</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     EggNOG_ID uniprot uniprot_gene_name\n",
       "0  zmo:ZMO1352  D2YW23           ZMO1352\n",
       "1  zmo:ZMO1351  Q5NMT5           ZMO1351\n",
       "2  zmo:ZMO1350  H2VFS2           ZMO1350\n",
       "3  zmo:ZMO1348  H2VFN6       map ZMO1348\n",
       "4  zmo:ZMO1347  H2VFP6           ZMO1347"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping_uniprot = pd.read_csv(os.path.join('..','data','external','uniprot_mapping.tsv'),sep = '\\t')\n",
    "mapping_uniprot = mapping_uniprot.rename(columns = {'Gene Names' : 'uniprot_gene_name', 'From':'EggNOG_ID','Entry':'uniprot'})[['EggNOG_ID','uniprot','uniprot_gene_name']]\n",
    "mapping_uniprot = mapping_uniprot.drop_duplicates('EggNOG_ID')\n",
    "mapping_uniprot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>ncbi_protein</th>\n",
       "      <th>kegg_maps</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>COG</th>\n",
       "      <th>EggNOG_ID</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot_gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_1768</th>\n",
       "      <td>CP006891.1</td>\n",
       "      <td>32.0</td>\n",
       "      <td>841.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Phage capsid scaffolding protein (GPO) serine ...</td>\n",
       "      <td>AHB11032.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td>702113.PP1Y_AT3685</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1853</th>\n",
       "      <td>CP006893.1</td>\n",
       "      <td>80.0</td>\n",
       "      <td>2122.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>outer membrane receptor protein</td>\n",
       "      <td>AHB11110.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>P</td>\n",
       "      <td>555217.Zmob_1838</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1912</th>\n",
       "      <td>CP006895.1</td>\n",
       "      <td>450.0</td>\n",
       "      <td>680.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11159.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>-</td>\n",
       "      <td>1296990.H845_3278</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1913</th>\n",
       "      <td>CP006895.1</td>\n",
       "      <td>677.0</td>\n",
       "      <td>1057.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>prophage maintenance system killer protein</td>\n",
       "      <td>AHB11160.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td>107635.AZUO01000001_gene1405</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1769</th>\n",
       "      <td>CP006891.1</td>\n",
       "      <td>974.0</td>\n",
       "      <td>2749.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11033.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>S</td>\n",
       "      <td>1522072.IL54_0955</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession  start     end strand gene_name old_locus_tag  \\\n",
       "locus_tag                                                             \n",
       "ZCP4_1768  CP006891.1   32.0   841.0      -      None          None   \n",
       "ZCP4_1853  CP006893.1   80.0  2122.0      -      None          None   \n",
       "ZCP4_1912  CP006895.1  450.0   680.0      +      None          None   \n",
       "ZCP4_1913  CP006895.1  677.0  1057.0      +      None          None   \n",
       "ZCP4_1769  CP006891.1  974.0  2749.0      +      None          None   \n",
       "\n",
       "                                                gene_product ncbi_protein  \\\n",
       "locus_tag                                                                   \n",
       "ZCP4_1768  Phage capsid scaffolding protein (GPO) serine ...   AHB11032.1   \n",
       "ZCP4_1853                    outer membrane receptor protein   AHB11110.1   \n",
       "ZCP4_1912                               hypothetical protein   AHB11159.1   \n",
       "ZCP4_1913         prophage maintenance system killer protein   AHB11160.1   \n",
       "ZCP4_1769                               hypothetical protein   AHB11033.1   \n",
       "\n",
       "          kegg_maps EC_number COG                     EggNOG_ID uniprot  \\\n",
       "locus_tag                                                                 \n",
       "ZCP4_1768                       S            702113.PP1Y_AT3685     NaN   \n",
       "ZCP4_1853                       P              555217.Zmob_1838     NaN   \n",
       "ZCP4_1912                       -             1296990.H845_3278     NaN   \n",
       "ZCP4_1913                       S  107635.AZUO01000001_gene1405     NaN   \n",
       "ZCP4_1769                       S             1522072.IL54_0955     NaN   \n",
       "\n",
       "          uniprot_gene_name  \n",
       "locus_tag                    \n",
       "ZCP4_1768               NaN  \n",
       "ZCP4_1853               NaN  \n",
       "ZCP4_1912               NaN  \n",
       "ZCP4_1913               NaN  \n",
       "ZCP4_1769               NaN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge with current annotation\n",
    "DF_annot = pd.merge(DF_annot.reset_index(),mapping_uniprot,how='left',on='EggNOG_ID')\n",
    "DF_annot.set_index('locus_tag',inplace=True)\n",
    "gene_names = []\n",
    "for index, row in DF_annot.iterrows():\n",
    "    gene_name = None\n",
    "    if row['gene_name'] == None:\n",
    "        uni_name = str(row['uniprot_gene_name'])\n",
    "        if uni_name != 'nan' and ' ' in uni_name:\n",
    "            gene_name = uni_name.split(' ')[0]\n",
    "            if 'zmo' in gene_name.lower():\n",
    "                gene_name = None\n",
    "    gene_names.append(gene_name) \n",
    "DF_annot['gene_name'] = gene_names\n",
    "DF_annot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>old_locus_tag</th>\n",
       "      <th>gene_product</th>\n",
       "      <th>ncbi_protein</th>\n",
       "      <th>kegg_maps</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>COG</th>\n",
       "      <th>EggNOG_ID</th>\n",
       "      <th>uniprot</th>\n",
       "      <th>uniprot_gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_0386</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>425778.0</td>\n",
       "      <td>426584.0</td>\n",
       "      <td>+</td>\n",
       "      <td>aat</td>\n",
       "      <td>None</td>\n",
       "      <td>leucyl/phenylalanyl-tRNA--protein transferase</td>\n",
       "      <td>AHB09706.1</td>\n",
       "      <td></td>\n",
       "      <td>2.3.2.6</td>\n",
       "      <td>O</td>\n",
       "      <td>zmo:ZMO0927</td>\n",
       "      <td>Q5NP09</td>\n",
       "      <td>aat ZMO0927</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0684</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>776548.0</td>\n",
       "      <td>777489.0</td>\n",
       "      <td>-</td>\n",
       "      <td>accA</td>\n",
       "      <td>None</td>\n",
       "      <td>acetyl-CoA carboxylase carboxyltransferase sub...</td>\n",
       "      <td>AHB09993.1</td>\n",
       "      <td>map00061;map00620;map00640;map00720;map01100;m...</td>\n",
       "      <td>2.1.3.15,6.4.1.2</td>\n",
       "      <td>I</td>\n",
       "      <td>zmo:ZMO0599</td>\n",
       "      <td>Q5NPY2</td>\n",
       "      <td>accA ZMO0599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0699</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>789224.0</td>\n",
       "      <td>790078.0</td>\n",
       "      <td>+</td>\n",
       "      <td>accD</td>\n",
       "      <td>None</td>\n",
       "      <td>acetyl-CoA carboxylase carboxyltransferase sub...</td>\n",
       "      <td>AHB10008.1</td>\n",
       "      <td>map00061;map00620;map00640;map00720;map01100;m...</td>\n",
       "      <td>2.1.3.15,6.4.1.2</td>\n",
       "      <td>I</td>\n",
       "      <td>zmo:ZMO0583</td>\n",
       "      <td>Q5NPZ8</td>\n",
       "      <td>accD ZMO0583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_0061</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>70267.0</td>\n",
       "      <td>70503.0</td>\n",
       "      <td>+</td>\n",
       "      <td>acpP</td>\n",
       "      <td>None</td>\n",
       "      <td>acyl carrier protein</td>\n",
       "      <td>AHB09399.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>I</td>\n",
       "      <td>zmo:ZMO1279</td>\n",
       "      <td>Q5NN07</td>\n",
       "      <td>acpP ZMO1279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1454</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1627254.0</td>\n",
       "      <td>1627658.0</td>\n",
       "      <td>-</td>\n",
       "      <td>acpS</td>\n",
       "      <td>None</td>\n",
       "      <td>holo-(acyl-carrier-protein) synthase</td>\n",
       "      <td>AHB10735.1</td>\n",
       "      <td>map00770</td>\n",
       "      <td>2.7.8.7</td>\n",
       "      <td>I</td>\n",
       "      <td>zmo:ZMO1709</td>\n",
       "      <td>Q5NLS7</td>\n",
       "      <td>acpS ZMO1709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1760</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1990403.0</td>\n",
       "      <td>1993540.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>diguanylate cyclase (GGDEF) domain-containing ...</td>\n",
       "      <td>AHB11025.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>zmo:ZMO1365</td>\n",
       "      <td>Q5NMS1</td>\n",
       "      <td>ZMO1365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1761</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1993874.0</td>\n",
       "      <td>1995235.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>oxygen-independent coproporphyrinogen III oxidase</td>\n",
       "      <td>AHB11026.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>H</td>\n",
       "      <td>zmo:ZMO1364</td>\n",
       "      <td>Q5NMS2</td>\n",
       "      <td>ZMO1364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1762</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1995263.0</td>\n",
       "      <td>1995484.0</td>\n",
       "      <td>-</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11027.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1763</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1995787.0</td>\n",
       "      <td>1995936.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11028.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1764</th>\n",
       "      <td>CP006818.1</td>\n",
       "      <td>1995926.0</td>\n",
       "      <td>1996108.0</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>AHB11029.1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1836 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession      start        end strand gene_name old_locus_tag  \\\n",
       "locus_tag                                                                    \n",
       "ZCP4_0386  CP006818.1   425778.0   426584.0      +       aat          None   \n",
       "ZCP4_0684  CP006818.1   776548.0   777489.0      -      accA          None   \n",
       "ZCP4_0699  CP006818.1   789224.0   790078.0      +      accD          None   \n",
       "ZCP4_0061  CP006818.1    70267.0    70503.0      +      acpP          None   \n",
       "ZCP4_1454  CP006818.1  1627254.0  1627658.0      -      acpS          None   \n",
       "...               ...        ...        ...    ...       ...           ...   \n",
       "ZCP4_1760  CP006818.1  1990403.0  1993540.0      -      None          None   \n",
       "ZCP4_1761  CP006818.1  1993874.0  1995235.0      -      None          None   \n",
       "ZCP4_1762  CP006818.1  1995263.0  1995484.0      -      None          None   \n",
       "ZCP4_1763  CP006818.1  1995787.0  1995936.0      +      None          None   \n",
       "ZCP4_1764  CP006818.1  1995926.0  1996108.0      +      None          None   \n",
       "\n",
       "                                                gene_product ncbi_protein  \\\n",
       "locus_tag                                                                   \n",
       "ZCP4_0386      leucyl/phenylalanyl-tRNA--protein transferase   AHB09706.1   \n",
       "ZCP4_0684  acetyl-CoA carboxylase carboxyltransferase sub...   AHB09993.1   \n",
       "ZCP4_0699  acetyl-CoA carboxylase carboxyltransferase sub...   AHB10008.1   \n",
       "ZCP4_0061                               acyl carrier protein   AHB09399.1   \n",
       "ZCP4_1454               holo-(acyl-carrier-protein) synthase   AHB10735.1   \n",
       "...                                                      ...          ...   \n",
       "ZCP4_1760  diguanylate cyclase (GGDEF) domain-containing ...   AHB11025.1   \n",
       "ZCP4_1761  oxygen-independent coproporphyrinogen III oxidase   AHB11026.1   \n",
       "ZCP4_1762                               hypothetical protein   AHB11027.1   \n",
       "ZCP4_1763                               hypothetical protein   AHB11028.1   \n",
       "ZCP4_1764                               hypothetical protein   AHB11029.1   \n",
       "\n",
       "                                                   kegg_maps  \\\n",
       "locus_tag                                                      \n",
       "ZCP4_0386                                                      \n",
       "ZCP4_0684  map00061;map00620;map00640;map00720;map01100;m...   \n",
       "ZCP4_0699  map00061;map00620;map00640;map00720;map01100;m...   \n",
       "ZCP4_0061                                                      \n",
       "ZCP4_1454                                           map00770   \n",
       "...                                                      ...   \n",
       "ZCP4_1760                                                      \n",
       "ZCP4_1761                                                      \n",
       "ZCP4_1762                                                      \n",
       "ZCP4_1763                                                      \n",
       "ZCP4_1764                                                      \n",
       "\n",
       "                  EC_number  COG    EggNOG_ID uniprot uniprot_gene_name  \n",
       "locus_tag                                                                \n",
       "ZCP4_0386           2.3.2.6    O  zmo:ZMO0927  Q5NP09       aat ZMO0927  \n",
       "ZCP4_0684  2.1.3.15,6.4.1.2    I  zmo:ZMO0599  Q5NPY2      accA ZMO0599  \n",
       "ZCP4_0699  2.1.3.15,6.4.1.2    I  zmo:ZMO0583  Q5NPZ8      accD ZMO0583  \n",
       "ZCP4_0061                      I  zmo:ZMO1279  Q5NN07      acpP ZMO1279  \n",
       "ZCP4_1454           2.7.8.7    I  zmo:ZMO1709  Q5NLS7      acpS ZMO1709  \n",
       "...                     ...  ...          ...     ...               ...  \n",
       "ZCP4_1760                      T  zmo:ZMO1365  Q5NMS1           ZMO1365  \n",
       "ZCP4_1761                      H  zmo:ZMO1364  Q5NMS2           ZMO1364  \n",
       "ZCP4_1762                    NaN                  NaN               NaN  \n",
       "ZCP4_1763                    NaN                  NaN               NaN  \n",
       "ZCP4_1764                    NaN                  NaN               NaN  \n",
       "\n",
       "[1836 rows x 14 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DF_annot.sort_values(by = 'gene_name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add Biocyc Operon information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To obtain operon information from Biocyc, follow the steps below\n",
    "\n",
    "1. Go to [Biocyc.org](https://biocyc.org/) (you may need to create an account and/or login)\n",
    "2. Change the organism database to your organism/strain\n",
    "3. Select **SmartTables** -> **Special SmartTables**\n",
    "4. Select **\"All genes of \\<organism\\>\"**\n",
    "5. Select the **\"Gene Name\"** column\n",
    "6. Under **\"ADD TRANSFORM COLUMN\"** select **\"Genes in same transcription unit\"**\n",
    "7. Select the **\"Genes in same transcription unit\"** column\n",
    "8. Under **\"ADD PROPERTY COLUMN\"** select **\"Accession-1\"**\n",
    "9. Under **OPERATIONS**, select **\"Export\"** -> **\"to Spreadsheet File...\"**\n",
    "10. Select **\"common names\"** and click **\"Export smarttable\"**\n",
    "11. Add file location below and run the code cell"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Left-End-Position</th>\n",
       "      <th>Right-End-Position</th>\n",
       "      <th>Product</th>\n",
       "      <th>Genes in same transcription unit</th>\n",
       "      <th>Accession-1.1</th>\n",
       "      <th>operon_list</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_1768</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1853</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1912</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1913</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1769</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1769</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene Name  Left-End-Position  Right-End-Position Product  \\\n",
       "locus_tag                                                            \n",
       "ZCP4_1768       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1853       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1912       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1913       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1769       NaN                NaN                 NaN     NaN   \n",
       "\n",
       "          Genes in same transcription unit Accession-1.1 operon_list  \n",
       "locus_tag                                                             \n",
       "ZCP4_1768                              NaN           NaN   ZCP4_1768  \n",
       "ZCP4_1853                              NaN           NaN   ZCP4_1853  \n",
       "ZCP4_1912                              NaN           NaN   ZCP4_1912  \n",
       "ZCP4_1913                              NaN           NaN   ZCP4_1913  \n",
       "ZCP4_1769                              NaN           NaN   ZCP4_1769  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biocyc_file = os.path.join('..','data','external','biocyc_annotations.txt')\n",
    "\n",
    "DF_biocyc = pd.read_csv(biocyc_file,sep='\\t')\n",
    "\n",
    "# Remove genes with no accession\n",
    "DF_biocyc = DF_biocyc[DF_biocyc['Accession-1'].notnull()]\n",
    "\n",
    "# Set the accession (i.e. locus tag) as index\n",
    "DF_biocyc = DF_biocyc.set_index('Accession-1').sort_values('Left-End-Position')\n",
    "\n",
    "# Specific for B. subtilis: Fix locus tags\n",
    "DF_biocyc.index = DF_biocyc.index.str.replace('BSU','BSU_')\n",
    "\n",
    "# Only keep genes in the final annotation file\n",
    "DF_biocyc = DF_biocyc.reindex(DF_annot.index)\n",
    "\n",
    "# Reformat transcription units\n",
    "DF_biocyc['operon_list'] = DF_biocyc['Accession-1.1'].apply(reformat_biocyc_tu)\n",
    "\n",
    "# Fill None with locus tags\n",
    "DF_biocyc['operon_list'].fillna(DF_biocyc.index.to_series(), inplace=True)\n",
    "\n",
    "DF_biocyc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign unique IDs to operons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following code assigns unique names to each operon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>Left-End-Position</th>\n",
       "      <th>Right-End-Position</th>\n",
       "      <th>Product</th>\n",
       "      <th>Genes in same transcription unit</th>\n",
       "      <th>Accession-1.1</th>\n",
       "      <th>operon_list</th>\n",
       "      <th>operon</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>locus_tag</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ZCP4_1768</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1768</td>\n",
       "      <td>Op0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1853</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1853</td>\n",
       "      <td>Op1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1912</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1912</td>\n",
       "      <td>Op2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1913</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1913</td>\n",
       "      <td>Op3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZCP4_1769</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZCP4_1769</td>\n",
       "      <td>Op4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Gene Name  Left-End-Position  Right-End-Position Product  \\\n",
       "locus_tag                                                            \n",
       "ZCP4_1768       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1853       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1912       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1913       NaN                NaN                 NaN     NaN   \n",
       "ZCP4_1769       NaN                NaN                 NaN     NaN   \n",
       "\n",
       "          Genes in same transcription unit Accession-1.1 operon_list operon  \n",
       "locus_tag                                                                    \n",
       "ZCP4_1768                              NaN           NaN   ZCP4_1768    Op0  \n",
       "ZCP4_1853                              NaN           NaN   ZCP4_1853    Op1  \n",
       "ZCP4_1912                              NaN           NaN   ZCP4_1912    Op2  \n",
       "ZCP4_1913                              NaN           NaN   ZCP4_1913    Op3  \n",
       "ZCP4_1769                              NaN           NaN   ZCP4_1769    Op4  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all operons\n",
    "operons = DF_biocyc['operon_list'].unique()\n",
    "\n",
    "# Map each operon to a unique string\n",
    "operon_dict = {operon: \"Op\"+str(i) for i, operon in enumerate(operons)}\n",
    "\n",
    "# Add names to dataframe\n",
    "DF_biocyc['operon'] = [operon_dict[op] for op in DF_biocyc[\"operon_list\"]]\n",
    "\n",
    "DF_biocyc.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, merge the Biocyc information with the main annotation DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_annot['operon'] = DF_biocyc['operon']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## regulator info from https://pmc.ncbi.nlm.nih.gov/articles/PMC7566282/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'fake' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1392/160509739.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mfake\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfake\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'fake' is not defined"
     ]
    }
   ],
   "source": [
    "fake = fake"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1054"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SI1 = pd.read_excel('../data/external/10_1128mSystems_00250-20/mSystems.00250-20-sd001.xlsx', sheet_name = 'TSSs')\n",
    "SI2 = pd.read_excel('../data/external/10_1128mSystems_00250-20/mSystems.00250-20-sd002.xlsx', sheet_name = 'Z. mobilis')\n",
    "rpoD_model_regulated = SI2.TSS.values\n",
    "true_false = []\n",
    "for TSS in SI1.Name:\n",
    "    true_false.append(TSS in rpoD_model_regulated)\n",
    "genes = set(SI1[true_false].Gene)\n",
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Name</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Type</th>\n",
       "      <th>A</th>\n",
       "      <th>B</th>\n",
       "      <th>C</th>\n",
       "      <th>D</th>\n",
       "      <th>E</th>\n",
       "      <th>F</th>\n",
       "      <th>A_Cov</th>\n",
       "      <th>B_Cov</th>\n",
       "      <th>C_Cov</th>\n",
       "      <th>D_Cov</th>\n",
       "      <th>E_Cov</th>\n",
       "      <th>F_Cov</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Class</th>\n",
       "      <th>Relative Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZM4</td>\n",
       "      <td>1463</td>\n",
       "      <td>TSS_0001</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>13.478958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>secondary TSS; excluded from analysis</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ZM4</td>\n",
       "      <td>1464</td>\n",
       "      <td>TSS_0002</td>\n",
       "      <td>-</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>44.787214</td>\n",
       "      <td>41.744175</td>\n",
       "      <td>23.488322</td>\n",
       "      <td>27.180245</td>\n",
       "      <td></td>\n",
       "      <td>18.142139</td>\n",
       "      <td>ZMO0001</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ZM4</td>\n",
       "      <td>3702</td>\n",
       "      <td>TSS_0003</td>\n",
       "      <td>+</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>19.954372</td>\n",
       "      <td>93.261609</td>\n",
       "      <td>13.737773</td>\n",
       "      <td>67.270941</td>\n",
       "      <td></td>\n",
       "      <td>39.560877</td>\n",
       "      <td>ZMO0003</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ZM4</td>\n",
       "      <td>6905</td>\n",
       "      <td>TSS_0004</td>\n",
       "      <td>+</td>\n",
       "      <td>P</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.36532</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ZMO0006</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ZM4</td>\n",
       "      <td>9461</td>\n",
       "      <td>TSS_0005</td>\n",
       "      <td>+</td>\n",
       "      <td>P</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.017413</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ZMO0008</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4647</th>\n",
       "      <td>pZM39</td>\n",
       "      <td>34729</td>\n",
       "      <td>TSS_4648</td>\n",
       "      <td>-</td>\n",
       "      <td>P</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.729131</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ZMOp39x030</td>\n",
       "      <td>TSS-Gene pair</td>\n",
       "      <td>-502.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4648</th>\n",
       "      <td>pZM39</td>\n",
       "      <td>35128</td>\n",
       "      <td>TSS_4649</td>\n",
       "      <td>-</td>\n",
       "      <td>P</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>74.443411</td>\n",
       "      <td>73.03457</td>\n",
       "      <td>11.105117</td>\n",
       "      <td>41.097151</td>\n",
       "      <td>ZMOp39x031</td>\n",
       "      <td>antisense</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4649</th>\n",
       "      <td>pZM39</td>\n",
       "      <td>38693</td>\n",
       "      <td>TSS_4650</td>\n",
       "      <td>-</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>29.052834</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>19.195156</td>\n",
       "      <td></td>\n",
       "      <td>ZMOp39x037</td>\n",
       "      <td>TSS-Gene pair</td>\n",
       "      <td>-158.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4650</th>\n",
       "      <td>pZM39</td>\n",
       "      <td>38695</td>\n",
       "      <td>TSS_4651</td>\n",
       "      <td>-</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>19.974643</td>\n",
       "      <td></td>\n",
       "      <td>64.300375</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>ZMOp39x037</td>\n",
       "      <td>TSS-Gene pair</td>\n",
       "      <td>-160.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4651</th>\n",
       "      <td>pZM39</td>\n",
       "      <td>38817</td>\n",
       "      <td>TSS_4652</td>\n",
       "      <td>+</td>\n",
       "      <td>P</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>342.090384</td>\n",
       "      <td></td>\n",
       "      <td>109.261914</td>\n",
       "      <td></td>\n",
       "      <td>224.138628</td>\n",
       "      <td>440.483815</td>\n",
       "      <td>ZMOp39x034</td>\n",
       "      <td>leaderless</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4652 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Sequence  Position      Name Strand Type  A  B  C  D  E  F       A_Cov  \\\n",
       "0         ZM4      1463  TSS_0001      -    S                 1               \n",
       "1         ZM4      1464  TSS_0002      -    P  1  1  1  1     1   44.787214   \n",
       "2         ZM4      3702  TSS_0003      +    P  1  1  1  1     1   19.954372   \n",
       "3         ZM4      6905  TSS_0004      +    P     1                           \n",
       "4         ZM4      9461  TSS_0005      +    P           1                     \n",
       "...       ...       ...       ...    ...  ... .. .. .. .. .. ..         ...   \n",
       "4647    pZM39     34729  TSS_4648      -    P           1                     \n",
       "4648    pZM39     35128  TSS_4649      -    P        1  1  1  1               \n",
       "4649    pZM39     38693  TSS_4650      -    P  1           1      29.052834   \n",
       "4650    pZM39     38695  TSS_4651      -    P  1     1            19.974643   \n",
       "4651    pZM39     38817  TSS_4652      +    P  1     1     1  1  342.090384   \n",
       "\n",
       "          B_Cov       C_Cov      D_Cov       E_Cov       F_Cov        Gene  \\\n",
       "0                                                    13.478958         NaN   \n",
       "1     41.744175   23.488322  27.180245               18.142139     ZMO0001   \n",
       "2     93.261609   13.737773  67.270941               39.560877     ZMO0003   \n",
       "3       7.36532                                                    ZMO0006   \n",
       "4                             7.017413                             ZMO0008   \n",
       "...         ...         ...        ...         ...         ...         ...   \n",
       "4647                          7.729131                          ZMOp39x030   \n",
       "4648              74.443411   73.03457   11.105117   41.097151  ZMOp39x031   \n",
       "4649                                     19.195156              ZMOp39x037   \n",
       "4650              64.300375                                     ZMOp39x037   \n",
       "4651             109.261914             224.138628  440.483815  ZMOp39x034   \n",
       "\n",
       "                                      Class  Relative Location  \n",
       "0     secondary TSS; excluded from analysis                NaN  \n",
       "1                                 antisense                NaN  \n",
       "2                                 antisense                NaN  \n",
       "3                                 antisense                NaN  \n",
       "4                                 antisense                NaN  \n",
       "...                                     ...                ...  \n",
       "4647                          TSS-Gene pair             -502.0  \n",
       "4648                              antisense                NaN  \n",
       "4649                          TSS-Gene pair             -158.0  \n",
       "4650                          TSS-Gene pair             -160.0  \n",
       "4651                             leaderless                0.0  \n",
       "\n",
       "[4652 rows x 20 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SI1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TSS</th>\n",
       "      <th>-35</th>\n",
       "      <th>Ri(-35)</th>\n",
       "      <th>Spacer (nt)</th>\n",
       "      <th>-10</th>\n",
       "      <th>Ri(-10)</th>\n",
       "      <th>GS</th>\n",
       "      <th>Ri(total)</th>\n",
       "      <th>Discriminator (nt)</th>\n",
       "      <th>Figure 5D</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSS_0002</td>\n",
       "      <td>TTGCCA</td>\n",
       "      <td>5.852980</td>\n",
       "      <td>19</td>\n",
       "      <td>GAAAGA</td>\n",
       "      <td>1.150175</td>\n",
       "      <td>2.742038</td>\n",
       "      <td>9.745193</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TSS_0003</td>\n",
       "      <td>TTCCCG</td>\n",
       "      <td>1.535923</td>\n",
       "      <td>17</td>\n",
       "      <td>TAGGAT</td>\n",
       "      <td>2.791982</td>\n",
       "      <td>2.202337</td>\n",
       "      <td>6.530242</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TSS_0004</td>\n",
       "      <td>TTGATT</td>\n",
       "      <td>6.507509</td>\n",
       "      <td>20</td>\n",
       "      <td>TAATCG</td>\n",
       "      <td>2.459706</td>\n",
       "      <td>3.535088</td>\n",
       "      <td>12.502302</td>\n",
       "      <td>5</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TSS_0006</td>\n",
       "      <td>TTTCAG</td>\n",
       "      <td>3.257833</td>\n",
       "      <td>17</td>\n",
       "      <td>TATAAC</td>\n",
       "      <td>6.007988</td>\n",
       "      <td>2.202337</td>\n",
       "      <td>11.468158</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TSS_0007</td>\n",
       "      <td>TGGCAA</td>\n",
       "      <td>2.723694</td>\n",
       "      <td>16</td>\n",
       "      <td>TATTAT</td>\n",
       "      <td>6.344320</td>\n",
       "      <td>2.330962</td>\n",
       "      <td>11.398977</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1786</th>\n",
       "      <td>TSS_4644</td>\n",
       "      <td>ATGCTT</td>\n",
       "      <td>0.997095</td>\n",
       "      <td>16</td>\n",
       "      <td>TAAATT</td>\n",
       "      <td>5.491830</td>\n",
       "      <td>2.330962</td>\n",
       "      <td>8.819887</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1787</th>\n",
       "      <td>TSS_4646</td>\n",
       "      <td>TTGCAG</td>\n",
       "      <td>4.784201</td>\n",
       "      <td>18</td>\n",
       "      <td>TATAAG</td>\n",
       "      <td>5.657147</td>\n",
       "      <td>2.330962</td>\n",
       "      <td>12.772310</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1788</th>\n",
       "      <td>TSS_4649</td>\n",
       "      <td>TTGTAT</td>\n",
       "      <td>5.666193</td>\n",
       "      <td>16</td>\n",
       "      <td>TATATG</td>\n",
       "      <td>5.060755</td>\n",
       "      <td>2.330962</td>\n",
       "      <td>13.057911</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1789</th>\n",
       "      <td>TSS_4651</td>\n",
       "      <td>TTTATA</td>\n",
       "      <td>5.145677</td>\n",
       "      <td>20</td>\n",
       "      <td>TAATAT</td>\n",
       "      <td>5.340058</td>\n",
       "      <td>3.535088</td>\n",
       "      <td>14.020822</td>\n",
       "      <td>8</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1790</th>\n",
       "      <td>TSS_4652</td>\n",
       "      <td>GTGTTT</td>\n",
       "      <td>0.122625</td>\n",
       "      <td>17</td>\n",
       "      <td>TAGAAT</td>\n",
       "      <td>4.331199</td>\n",
       "      <td>2.202337</td>\n",
       "      <td>6.656161</td>\n",
       "      <td>6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1791 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           TSS     -35   Ri(-35)  Spacer (nt)     -10   Ri(-10)        GS  \\\n",
       "0     TSS_0002  TTGCCA  5.852980           19  GAAAGA  1.150175  2.742038   \n",
       "1     TSS_0003  TTCCCG  1.535923           17  TAGGAT  2.791982  2.202337   \n",
       "2     TSS_0004  TTGATT  6.507509           20  TAATCG  2.459706  3.535088   \n",
       "3     TSS_0006  TTTCAG  3.257833           17  TATAAC  6.007988  2.202337   \n",
       "4     TSS_0007  TGGCAA  2.723694           16  TATTAT  6.344320  2.330962   \n",
       "...        ...     ...       ...          ...     ...       ...       ...   \n",
       "1786  TSS_4644  ATGCTT  0.997095           16  TAAATT  5.491830  2.330962   \n",
       "1787  TSS_4646  TTGCAG  4.784201           18  TATAAG  5.657147  2.330962   \n",
       "1788  TSS_4649  TTGTAT  5.666193           16  TATATG  5.060755  2.330962   \n",
       "1789  TSS_4651  TTTATA  5.145677           20  TAATAT  5.340058  3.535088   \n",
       "1790  TSS_4652  GTGTTT  0.122625           17  TAGAAT  4.331199  2.202337   \n",
       "\n",
       "      Ri(total)  Discriminator (nt) Figure 5D  \n",
       "0      9.745193                   4       NaN  \n",
       "1      6.530242                   6       NaN  \n",
       "2     12.502302                   5       NaN  \n",
       "3     11.468158                   6       NaN  \n",
       "4     11.398977                   6       NaN  \n",
       "...         ...                 ...       ...  \n",
       "1786   8.819887                   6       NaN  \n",
       "1787  12.772310                   6       NaN  \n",
       "1788  13.057911                   6       NaN  \n",
       "1789  14.020822                   8       NaN  \n",
       "1790   6.656161                   6       NaN  \n",
       "\n",
       "[1791 rows x 10 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SI2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Clean up and save annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we will re-order the annotation columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "order = ['gene_name','accession','start','end','strand','gene_product','COG','uniprot','operon', 'kegg_maps', 'EC_number']\n",
    "    \n",
    "DF_annot = DF_annot[order]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_annot.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following graphs show how much information is available for the organism."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sns.set_style('ticks')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax = plt.subplots()\n",
    "DF_annot.count().plot(kind='bar',ax=ax)\n",
    "ax.set_ylabel('# of Values',fontsize=18)\n",
    "ax.tick_params(labelsize=16)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fill missing values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some organisms are missing gene names, so these will be filled with locus tag gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Fill in missing gene names with locus tag names\n",
    "DF_annot['tmp_name'] = DF_annot.copy().index.tolist()\n",
    "DF_annot.gene_name.fillna(DF_annot.tmp_name,inplace=True)\n",
    "DF_annot.drop('tmp_name',axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " COG letters will also be converted to the full name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fill missing COGs with X\n",
    "DF_annot['COG'].fillna('X',inplace=True)\n",
    "DF_annot['COG'] = DF_annot['COG'].str.replace('-', 'X')\n",
    "\n",
    "# Change single letter COG annotation to full description\n",
    "DF_annot['COG'] = DF_annot.COG.apply(cog2str)\n",
    "\n",
    "counts = DF_annot.COG.value_counts()\n",
    "plt.pie(counts.values,labels=counts.index);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Uncomment the following line to save the gene annotation dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_annot.to_csv(os.path.join('..','data','processed_data','gene_info.csv'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO Annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To start, download the GO Annotations for your organism from AmiGO 2\n",
    "\n",
    "1. Go to [AmiGO 2](http://amigo.geneontology.org/amigo/search/annotation)\n",
    "1. Filter for your organism\n",
    "1. Click ``CustomDL``\n",
    "1. Drag ``GO class (direct)`` to the end of your Selected Fields\n",
    "1. Enter the location of your GO annotation file below and run the following code block"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "go_file = os.path.join('..','data','external','GO_annotations.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO = pd.read_csv(go_file,sep='\\t',header=None,usecols=[2,17])\n",
    "DF_GO.columns = ['gene_name','gene_ontology']\n",
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Convert the gene names to gene locus tags, and drop gene names that cannot be converted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "name2num = {v:k for k,v in DF_annot.gene_name.to_dict().items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO['gene_id'] = [name2num[x] if x in name2num.keys() else None for x in DF_GO.gene_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we remove null entries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO = DF_GO[DF_GO.gene_id.notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DF_GO.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Uncomment the line below to save the annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "DF_GO[['gene_id','gene_name','gene_ontology']].to_csv(os.path.join('..','data','external','GO_annotations_curated.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.16"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
